_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_splitter.cpp

```diff
@@ -0,0 +1,259 @@
+
+        if (m_cache.size())
+        {
+            nalu = { m_cache.data(), m_cache.data() + m_cache.size(), m_cache[prefix_size], m_pts } ; // Construct unit from cached data
+            m_needCleanCache = true; // Need to clean up the cache on next iteration
+        }
+
+        return nalu;
+    }
+
+    uint8_t * RawHeaderIterator::FindStartCode(uint8_t * begin, uint8_t * end)
+    {
+        for (; begin < end - 3; ++begin)
+        {
+            if (begin[0] == 0 && begin[1] == 0 && begin[2] == 1)
+            {
+                return begin;
+            }
+        }
+        return nullptr;
+    }
+
+    RawUnit RawHeaderIterator::FindRawUnit(mfxU8 * begin, mfxU8 * end)
+    {
+        for (; begin < end - 4; ++begin)
+        {
+            if (begin[0] == 0 && begin[1] == 0 && begin[2] == 1)
+            {
+                mfxU8 type = begin[3];
+
+                for (mfxU8 * next = begin + 4; next < end - 3; ++next)
+                {
+                    if (next[0] == 0 && next[1] == 0 && next[2] == 1)
+                    {
+                        return RawUnit(begin, next, type, -1);
+                    }
+                }
+
+                return RawUnit(begin, nullptr, type, -1);
+            }
+        }
+
+        return RawUnit();
+    }
+
+    RawUnit Splitter::GetUnits(UMC::MediaData* source)
+    {
+        if (m_cleanUpCache)
+        {
+            m_buff.clear();
+            m_cleanUpCache = false;
+        }
+
+        m_it.LoadData(source); // Find new unit
+
+        // for loop is to handle cases like: seq header, seq header, seq extension
+        for (;;)
+        {
+            if (m_buff.size() && (m_it != RawHeaderIterator()))
+            {
+                size_t size = m_it->end - m_it->begin;
+                MPEG2_UNIT_TYPE_EXT type_ext = (MPEG2_UNIT_TYPE_EXT) ((m_it->type == EXTENSION && size >= 5) ? (m_it->begin[prefix_size + 1] >> 4) : 0 );
+                if (m_typeExtToFind == type_ext)
+                {
+                    m_buff.insert(m_buff.end(), (uint8_t *)m_it->begin, (uint8_t *)m_it->end);
+
+                    m_cleanUpCache = true;
+                    m_typeExtToFind = 0;
+
+                    return { m_buff.data(), m_buff.data() + m_buff.size(), m_buff[prefix_size], -1 };
+                }
+                else
+                {
+                    m_buff.clear(); // drop sequence/picture header without corresponding sequence/picture extension header
+                    m_typeExtToFind = 0;
+                    continue;
+                }
+            }
+            else if (SEQUENCE_HEADER == m_it->type || PICTURE_HEADER == m_it->type)
+            {
+                m_typeExtToFind = (SEQUENCE_HEADER == m_it->type) ? SEQUENCE_EXTENSION : PICTURE_CODING_EXTENSION;
+                m_buff.insert(m_buff.end(), (uint8_t *)m_it->begin, (uint8_t *)m_it->end);
+
+                ++m_it; // Look for another unit
+
+                if (m_it != RawHeaderIterator())
```

Comment:
This piece of code (under condition) looks to be the same as above. Is it possible to extract this as function (lambda)?

Reply:
I'll think.